src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,287 @@
+
+lemma coe.add_monoid_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.add_monoid_hom R K) x v = x v := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` is a ring homomorphism. -/
+noncomputable! def coe.ring_hom : ring_hom (R_hat R K) (K_hat R K)  :=
+{ to_fun   := coe,
+  map_one' := rfl,
+  map_mul' := λ x y, by {ext p, simp only [pi.mul_apply, subring.coe_mul], refl },
+  ..coe.add_monoid_hom R K }
+
+lemma coe.ring_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.ring_hom R K) x v = x v := rfl
+
+end finite_integral_adeles
+
+section algebra_instances
+
+instance : algebra K (K_hat R K) :=
+(by apply_instance : algebra K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance prod_completions.algebra' : algebra R (K_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion K))
+
+instance : is_scalar_tower R K (K_hat R K) :=
+(by apply_instance : is_scalar_tower R K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance : algebra R (R_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion_integers K))
+
+instance prod_completions.algebra_completions : algebra (R_hat R K) (K_hat R K) :=
+(finite_integral_adeles.coe.ring_hom R K).to_algebra
+
+instance prod_completions.is_scalar_tower_completions :
+  is_scalar_tower R (R_hat R K) (K_hat R K) :=
+(by apply_instance : is_scalar_tower R (Π (v : height_one_spectrum R), v.adic_completion_integers K)
+   (Π (v : height_one_spectrum R), v.adic_completion K))
+
+end algebra_instances
+
+namespace finite_integral_adeles
+
+/-- The inclusion of `R_hat` in `K_hat` is a ring homomorphism. -/
+noncomputable! def coe.alg_hom : alg_hom R (R_hat R K) (K_hat R K)  :=
+{ to_fun    := coe,
+  commutes' := λ r, rfl,
+  ..coe.ring_hom R K  }
+
+lemma coe.alg_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.alg_hom R K) x v = x v := rfl
+
+end finite_integral_adeles
+
+/-! ### The finite adèle ring of a Dedekind domain
+We define the finite adèle ring of `R` as the restricted product over all maximal ideals `v` of `R`
+of `adic_completion` with respect to `adic_completion_integers`. We prove that it is a commutative
+ring. TODO: show that it is a topological ring with the restricted product topology. -/
+
+/-- The finite adèle ring of `R` is the restricted product over all maximal ideals `v` of `R`
+of `adic_completion` with respect to `adic_completion_integers`.-/
+def finite_adele_ring := { x : (K_hat R K) //
+  ∀ᶠ (v : height_one_spectrum R) in filter.cofinite, (x v ∈ v.adic_completion_integers K) }
+
+/-- The coercion map from `finite_adele_ring R K` to `K_hat R K`. -/
+noncomputable! def coe' : (finite_adele_ring R K) → K_hat R K := λ x, x.val
+instance has_coe' : has_coe (finite_adele_ring R K) (K_hat R K) := {coe := coe' R K }
+instance has_lift_t' : has_lift_t (finite_adele_ring R K) (K_hat R K) := {lift := coe' R K }
```

Comment:
I'm not an expert in coercions but do you definitely need both `has_coe` and `has_lift_t`?

Reply:
I have removed the `has_lift_t` one.